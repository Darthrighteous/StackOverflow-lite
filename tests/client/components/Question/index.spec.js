import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import ConnectedQuestion, { Question } from '../../../../client/components/Question';
import { mockSingleQuestionResponse } from '../../../support/mockData';

/**
 * setup for test
 * @returns {object} containing props and enzyme wrappers
 */
function setup() {
  const props = {
    isLoggedIn: true,
    match: {
      params: { id: 1 }
    },
    history: {},
    question: mockSingleQuestionResponse.question,
    comments: [],
    answers: [],
    fetchSingle: jest.fn(),
    postComment: jest.fn(),
    postAnswer: jest.fn(),
    modifyPost: jest.fn(),
    deletePost: jest.fn(),
  };
  const loggedOutProps = {
    ...props, isLoggedIn: false,
  };
  const withAnswersProps = {
    ...props, answers: mockSingleQuestionResponse.answers,
  };

  const mockStore = configureMockStore([]);
  const store = mockStore({
    global: {
      isLoggedIn: true
    },
    questions: [],
    singleQuestion: {
      question: {},
      comments: [],
      answers: [],
    },
  });

  localStorage.setItem('user', JSON.stringify({username: 'Darthrighteous'}));
  const questionWrapper = shallow(<Question {...props} />);
  localStorage.clear();
  const loggedOutWrapper =shallow(<Question {...loggedOutProps} />);
  const withAnswersWrapper =shallow(<Question {...withAnswersProps} />);
  const connectedQuestion = shallow(
    <ConnectedQuestion {...props} store={store} />
  );

  return {
    questionWrapper,
    loggedOutWrapper,
    withAnswersWrapper,
    connectedQuestion,
  };
}

describe('Question Component Test', () => {
  const {
    questionWrapper, loggedOutWrapper, withAnswersWrapper, connectedQuestion
  } = setup();

  it('should match snapshot', () => {
    expect(questionWrapper).toMatchSnapshot();
    expect(loggedOutWrapper).toMatchSnapshot();
    expect(withAnswersWrapper).toMatchSnapshot();
    expect(connectedQuestion).toMatchSnapshot();
  });

  test('set sort handler', () => {
    const instance = questionWrapper.instance();
    instance.setSort('test sort');
    expect(instance.state.sort).toBe('test sort');
    instance.setSort('test sort');
    expect(instance.state.sort).toBe('test sort');
  });

  test('answer input handler', () => {
    const event = {
      target: { value: 'testanswer' },
    };

    questionWrapper.find('.answer-body').simulate('change', event);
    expect(questionWrapper.state().answerInput).toEqual('testanswer');
  });

  test('handle comment post', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        querySelector: () => ({ value: 'test'})
      }
    };

    const instance = questionWrapper.instance();
    instance.handleCommentPost('questions', 1, event);
  });

  test('handle answer post', () => {
    questionWrapper.find('.btn-post-answer').simulate('click');
    
  });

  test('handle accept click', () => {
    withAnswersWrapper.find('Post').at(2).dive().find('.btn-accept').simulate('click');
  });

  test('handle delete click', () => {
    const instance = questionWrapper.instance();
    instance.handleDeletePost('questions', null);
  });
});
