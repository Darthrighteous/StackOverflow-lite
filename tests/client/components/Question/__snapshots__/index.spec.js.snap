// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component Test should match snapshot 1`] = `
<div
  className="main-content"
>
  <div
    className="content-header"
  >
    <h2>
      Why do we celebrate when we win?
    </h2>
    <ButtonNewQuestion />
  </div>
  <div
    className="question-answers-container"
  >
    <div
      className="question-body"
    >
      <Post
        comments={Array []}
        isLoggedIn={true}
        isQuestionOwner={true}
        onCommentPost={[Function]}
        onDeletePost={[Function]}
        onModifyPost={[Function]}
        post={
          Object {
            "accepted_answer": null,
            "answer_count": 3,
            "body": "What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?",
            "comment_count": 8,
            "created_at": "2018-12-26T08:26:53.488Z",
            "id": 4,
            "score": 0,
            "title": "Why do we celebrate when we win?",
            "username": "Darthrighteous",
          }
        }
        type="questions"
      />
    </div>
    <ContentOptions
      background={true}
      isAnswerDropdown={true}
      onSortClick={[Function]}
      sort="most recent"
      title="Answers"
    />
    <div
      className="answers"
    >
      <div
        className="answers-list"
      >
        <div
          className="no-resource-message"
        >
          <span>
            No answers yet, be the first one to post an answer
          </span>
        </div>
      </div>
      <div
        className="new-answer"
      >
        <h3>
          Your Answer
        </h3>
        <textarea
          className="answer-body"
          onChange={[Function]}
          placeholder="Type your answer here..."
          rows="10"
          value=""
        />
        <button
          className="btn-post-answer"
          onClick={[Function]}
          type="button"
        >
          Post Answer
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component Test should match snapshot 2`] = `
<div
  className="main-content"
>
  <div
    className="content-header"
  >
    <h2>
      Why do we celebrate when we win?
    </h2>
    <ButtonNewQuestion />
  </div>
  <div
    className="question-answers-container"
  >
    <div
      className="question-body"
    >
      <Post
        comments={Array []}
        isLoggedIn={false}
        isQuestionOwner={false}
        onCommentPost={[Function]}
        onDeletePost={[Function]}
        onModifyPost={[Function]}
        post={
          Object {
            "accepted_answer": null,
            "answer_count": 3,
            "body": "What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?",
            "comment_count": 8,
            "created_at": "2018-12-26T08:26:53.488Z",
            "id": 4,
            "score": 0,
            "title": "Why do we celebrate when we win?",
            "username": "Darthrighteous",
          }
        }
        type="questions"
      />
    </div>
    <ContentOptions
      background={true}
      isAnswerDropdown={true}
      onSortClick={[Function]}
      sort="most recent"
      title="Answers"
    />
    <div
      className="answers"
    >
      <div
        className="answers-list"
      >
        <div
          className="no-resource-message"
        >
          <span>
            No answers yet, be the first one to post an answer
          </span>
        </div>
      </div>
      <div
        className="no-answers-message"
      >
        <span>
          Please 
          <Link
            replace={false}
            to="/login"
          >
            login
          </Link>
           or 
          <Link
            replace={false}
            to="/signup"
          >
            signup
          </Link>
           to post an answer.
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component Test should match snapshot 3`] = `
<div
  className="main-content"
>
  <div
    className="content-header"
  >
    <h2>
      Why do we celebrate when we win?
    </h2>
    <ButtonNewQuestion />
  </div>
  <div
    className="question-answers-container"
  >
    <div
      className="question-body"
    >
      <Post
        comments={Array []}
        isLoggedIn={true}
        isQuestionOwner={false}
        onCommentPost={[Function]}
        onDeletePost={[Function]}
        onModifyPost={[Function]}
        post={
          Object {
            "accepted_answer": null,
            "answer_count": 3,
            "body": "What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?",
            "comment_count": 8,
            "created_at": "2018-12-26T08:26:53.488Z",
            "id": 4,
            "score": 0,
            "title": "Why do we celebrate when we win?",
            "username": "Darthrighteous",
          }
        }
        type="questions"
      />
    </div>
    <ContentOptions
      background={true}
      isAnswerDropdown={true}
      onSortClick={[Function]}
      sort="most recent"
      title="Answers"
    />
    <div
      className="answers"
    >
      <div
        className="answers-list"
      >
        <Post
          comments={
            Array [
              Object {
                "body": "test comment",
                "created_at": "cheeeee",
                "id": 1,
                "username": "tester",
              },
            ]
          }
          isLoggedIn={true}
          isQuestionOwner={false}
          key="5"
          onCommentPost={[Function]}
          onDeletePost={[Function]}
          onModifyPost={[Function]}
          post={
            Object {
              "accepted": false,
              "body": "Well, we celebrate winning cause it makes us feel good. You shouldn't worry about why we do it. All that matters is your happiness whether you've been programmed to do it or not.",
              "comment_count": 0,
              "commentlist": Array [
                Object {
                  "body": "test comment",
                  "created_at": "cheeeee",
                  "id": 1,
                  "username": "tester",
                },
              ],
              "created_at": "2018-12-29T13:15:20.577Z",
              "id": 5,
              "question_id": 4,
              "score": 0,
              "username": "Darthrighteous",
            }
          }
          type="answers"
        />
        <Post
          comments={Array []}
          isLoggedIn={true}
          isQuestionOwner={false}
          key="6"
          onCommentPost={[Function]}
          onDeletePost={[Function]}
          onModifyPost={[Function]}
          post={
            Object {
              "accepted": true,
              "body": "Duplicated. Well, we celebrate winning cause it makes us feel good. You shouldn't worry about why we do it. All that matters is your happiness whether you've been programmed to do it or not.",
              "comment_count": 0,
              "commentlist": Array [
                Object {
                  "body": null,
                  "created_at": null,
                  "id": null,
                  "username": null,
                },
              ],
              "created_at": "2018-12-29T13:16:51.334Z",
              "id": 6,
              "question_id": 4,
              "score": 0,
              "username": "Darthrighteous",
            }
          }
          type="answers"
        />
      </div>
      <div
        className="new-answer"
      >
        <h3>
          Your Answer
        </h3>
        <textarea
          className="answer-body"
          onChange={[Function]}
          placeholder="Type your answer here..."
          rows="10"
          value=""
        />
        <button
          className="btn-post-answer"
          onClick={[Function]}
          type="button"
        >
          Post Answer
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component Test should match snapshot 4`] = `
<Question
  answers={Array []}
  comments={Array []}
  deletePost={[Function]}
  fetchSingle={[Function]}
  history={Object {}}
  isLoggedIn={true}
  match={
    Object {
      "params": Object {
        "id": 1,
      },
    }
  }
  modifyPost={[Function]}
  postAnswer={[Function]}
  postComment={[Function]}
  question={Object {}}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
